### PR TITLE
[Feat]: SuperAdmin - Implement API endpoint to list registered users

### LINKED ISSUE #139: [FEAT]: Admin - User Listing: Implement API Endpoint to list registered users for admin
### Description
Create an API endpoint to retrieve all registered users in the application for the admin user. This endpoint will be part of the administration user management. Admin users can search users using this endpoint, and return all users using pagination

### Acceptance Criteria
- The API endpoint (/api/v1/admin/users) handles the GET request correctly and returns the expected response format.
- The endpoint returns a list of all users in the system using pagination to improve performance
- The endpoint should accommodate optional query parameters for filtering users by their first name, last name, and email
- This endpoint should be secured and only accessible to an admin user

### Purpose
To provide admin users with a comprehensive list of registered users, enabling effective navigation, and filtering.

### Requirements
- [ ] Create the API endpoint for retrieving registered users by an admin user
- [ ] Implement efficient database querying to fetch user data
- [ ] Implement optional query parameters for advanced use cases (limit, offset, email, first_name, last_name).
- [ ] Implement a caching mechanism to improve response times for frequent requests

**Endpoint**
`GET /api/v1/admin/users`

**Responses**
- Success 
Status Code: 200 OK

```
{
  "status_code": 200,
  "meta": {
    "total_pages": 0,
    "current_page": 0,
    "page_size": 0,
    "total_count": 0,
    "has_previous": true,
    "has_next": true
  },
  "users": [
    {
      "id": "3fa85f64-5717-4562-b3fc-2c963f66afa6",
      "first_name": "string",
      "last_name": "string",
      "email": "string",
      "avatar_url": 'string",
      "created_on": "2024-07-27T15:03:41.880Z"
    }
  ]
}
```

- Error 
Status Code: 400 Bad Request - Invalid query parameters
```
{
  "status_code": 400,
  "error": {
    "code": "INVALID_QUERY_PARAMETER",
    "message": "The provided query parameter is invalid.",
    "details": {
      "invalid_parameter": "limit",
      "reason": "Must be a positive integer"
    }
  }
}
```
Status Code: 500 Internal Server Error - Server-side error
```
{
 "status_code": 500,
 "error": {
    "code": "INTERNAL_SERVER_ERROR",
    "message": "An unexpected error occurred while processing your request.",
    "details": {
      "support_email": "support@example.com"
    }
  }
}
```

### Expected Outcome
- Admin users can retrieve a list of paginated registered users.
- The endpoint provides accurate and up-to-date user information
- The system handles errors gracefully, providing clear messages to the admin user.
- The endpoint can filter data based on optional query parameters

### Testing
- Perform manual tests to verify the API endpoint response structure and data accuracy.
- Write unit tests for the admin users listing retrieval logic.
- Write integration tests to ensure proper database interaction and response formatting.
- Conduct performance testing to verify response times under various loads.
- Test error scenarios to ensure proper error handling and messaging.

### PR DESCRIPTION
**Pull request by dmendie**
# Description

Feature list registered users endpoint for [139](https://github.com/hngprojects/hng_boilerplate_csharp_web/issues/139) 

Closes #139

# Changes proposed

## What were you told to do?
[FEAT]: Admin - User Listing: Implement API Endpoint to list registered users for admin

## What did you do?

- Added admin controller with GET endpoint for search users
- implement service logic for searching users
- Unit tests

# Check List (Check all the applicable boxes)

🚨Please review the [contribution guideline](CONTRIBUTING.md) for this repository.

- [x] My code follows the code style of this project.
- [x] This PR does not contain plagiarized content.
- [x] The title and description of the PR is clear and explains the approach.
- [x] I am making a pull request against the **dev branch** (left side).
- [x] My commit messages styles matches our requested structure.
- [x] My code additions will fail neither code linting checks nor unit test.
- [x] I am only making changes to files I was requested to.

# Screenshots/Videos

### CODE REVIEW

File: src/Hng.Application/Features/SuperAdmin/Dto/UserDto.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+﻿using System.Text.Json.Serialization;
+
+namespace Hng.Application.Features.SuperAdmin.Dto
+{
+    public class UserDto
+    {
+        public Guid Id { get; set; }
```
Comment (chimaihueze) [inline src/Hng.Application/Features/SuperAdmin/Dto/UserDto.cs:7]: Add Json property name for id too

File: src/Hng.Web/Controllers/AdminController.cs [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+        [ProducesResponseType(StatusCodes.Status200OK)]
+        public async Task<ActionResult> GetUsersBySearch([FromQuery] UsersQueryParameters parameters)
+        {
+            var users = await _mediator.Send(new GetUsersBySearchQuery(parameters));
+            return Ok(new
+            {
+                status_code = 200,
+                users.MetaData,
+                users
+            });
```
Comment (Godhanded) [inline src/Hng.Web/Controllers/AdminController.cs:34]: Please check the HNG application, there's a folder called shared, it has our generic response dto, return that and set the data field to this data you want to return

**Review by Godhanded — CHANGES_REQUESTED** [review]

File: src/Hng.Application/Features/SuperAdmin/Handlers/GetUsersBySearchQueryHandler.cs [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+using MediatR;
+
+namespace Hng.Application.Features.SuperAdmin.Handlers
+{
+    public class GetUsersBySearchQueryHandler : IRequestHandler<GetUsersBySearchQuery, PagedListDto<UserDto>>
+    {
+        private readonly IRepository<User> _userRepository;
+        private readonly IMapper _mapper;
+
+        public GetUsersBySearchQueryHandler(IRepository<User> userService, IMapper mapper)
```
Comment (chimaihueze) [inline src/Hng.Application/Features/SuperAdmin/Handlers/GetUsersBySearchQueryHandler.cs:16]: why userService? and not userRepository?

File: src/Hng.Application/Features/SuperAdmin/Dto/UserDto.cs (hunk 1 above)
Comment (chimaihueze) [inline src/Hng.Application/Features/SuperAdmin/Dto/UserDto.cs:7]: Add Json property name for id too

File: src/Hng.Application/Features/SuperAdmin/Handlers/GetUsersBySearchQueryHandler.cs (hunk 3 above)
Comment (chimaihueze) [inline src/Hng.Application/Features/SuperAdmin/Handlers/GetUsersBySearchQueryHandler.cs:16]: why userService? and not userRepository?

File: src/Hng.Application/Features/SuperAdmin/Mappers/AdminUsersMappingProfile.cs [hunk 4]
```diff
@@ -0,0 +1,13 @@
+﻿using Hng.Application.Features.SuperAdmin.Dto;
+using Hng.Domain.Entities;
+
+namespace Hng.Application.Features.SuperAdmin.Mappers
+{
+    public class AdminUsersMappingProfile : AutoMapper.Profile
```
Comment (chimaihueze) [inline src/Hng.Application/Features/SuperAdmin/Mappers/AdminUsersMappingProfile.cs:6]: what is the difference between this and the UserMappingProfile that already exists? It serves the same purpose.

File: src/Hng.Application/Shared/Dtos/PagedListMetadataDto.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
+﻿namespace Hng.Application.Shared.Dtos
+{
+    public class PagedListMetadataDto
+    {
+        public int TotalPages { get; set; }
+        public int CurrentPage { get; set; }
```
Comment (chimaihueze) [inline src/Hng.Application/Shared/Dtos/PagedListMetadataDto.cs:6]: use json property for snake case naming convention

**Review by chimaihueze — CHANGES_REQUESTED** [review]

File: src/Hng.Web/Controllers/AdminController.cs (hunk 2 above)
Comment (dmendie) [reply src/Hng.Web/Controllers/AdminController.cs:34]: The generic response dto does not satisfy the response data structure. this is what the response looks like when using the generic response dto

{
  "data": [
    {
      "id": "da037533-179f-434c-bb24-6ab966315672",
      "first_name": "Burdette",
      "last_name": "Grimes",
      "avatar_url": "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/33.jpg",
      "email": "Burdette0@hotmail.com",
      "created_on": "0001-01-01T00:00:00"
    },
    {
      "id": "f2755464-6326-47ad-9cff-7d5e62035599",
      "first_name": "Lemuel",
      "last_name": "Schneider",
      "avatar_url": "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1151.jpg",
      "email": "Lemuel.Schneider33@yahoo.com",
      "created_on": "0001-01-01T00:00:00"
    }
  ],
  "message": "Request completed successfully."
}

whereas, the response should look like this 

{
  "status_code": 200,
  "metaData": {
    "totalPages": 1,
    "currentPage": 1,
    "pageSize": 10,
    "totalCount": 2,
    "hasPrevious": false,
    "hasNext": false
  },
  "users": [
    {
      "id": "da037533-179f-434c-bb24-6ab966315672",
      "first_name": "Burdette",
      "last_name": "Grimes",
      "avatar_url": "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/33.jpg",
      "email": "Burdette0@hotmail.com",
      "created_on": "0001-01-01T00:00:00"
    },
    {
      "id": "f2755464-6326-47ad-9cff-7d5e62035599",
      "first_name": "Lemuel",
      "last_name": "Schneider",
      "avatar_url": "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1151.jpg",
      "email": "Lemuel.Schneider33@yahoo.com",
      "created_on": "0001-01-01T00:00:00"
    }
  ]
}

File: src/Hng.Application/Features/SuperAdmin/Mappers/AdminUsersMappingProfile.cs (hunk 4 above)
Comment (dmendie) [reply src/Hng.Application/Features/SuperAdmin/Mappers/AdminUsersMappingProfile.cs:6]: The userdto in user management contains fields that are not required on the endpoint like user organisations, products, profile etc, the admin user to requires the created_on property which the user management dto does not have. Extending an existing DTO may alter the behaviour or response of another endpoint

File: src/Hng.Web/Controllers/AdminController.cs (hunk 2 above)
Comment (Godhanded) [reply src/Hng.Web/Controllers/AdminController.cs:34]: Okay let's have this.
In the data for the generic response pass new{metadata, user}

File: src/Hng.Application/Features/SuperAdmin/Mappers/AdminUsersMappingProfile.cs (hunk 4 above)
Comment (chimaihueze) [reply src/Hng.Application/Features/SuperAdmin/Mappers/AdminUsersMappingProfile.cs:6]: > The userdto in user management contains fields that are not required on the endpoint like user organisations, products, profile etc, the admin user to requires the created_on property which the user management dto does not have. Extending an existing DTO may alter the behaviour or response of another endpoint

Okay, makes sense to me.

**Review by chimaihueze — APPROVED** [review]

File: src/Hng.Web/Controllers/AdminController.cs (hunk 2 above)
Comment (dmendie) [reply src/Hng.Web/Controllers/AdminController.cs:34]: Ok, what i'll do is create a generic response for paginated data

Comment (Godhanded) [reply src/Hng.Web/Controllers/AdminController.cs:34]: Sweet!

File: src/Hng.Application/Features/SuperAdmin/Handlers/GetUsersBySearchQueryHandler.cs
Comment (Godhanded) [inline src/Hng.Application/Features/SuperAdmin/Handlers/GetUsersBySearchQueryHandler.cs:1]: I think we would optimize this handler later, its going to be a bit heavy on memory

**Review by Godhanded — APPROVED** [review]

**Review by Ace2489 — APPROVED** [review]

**Pull request merged**